src/libraries/System.Private.CoreLib/src/System/Diagnostics/Debug.cs

```diff
@@ -82,10 +82,12 @@ public static int IndentSize
             Assert(condition, string.Empty, string.Empty);
 
         [Conditional("DEBUG")]
-        public static void Assert([DoesNotReturnIf(false)] bool condition, string? message) =>
+        [OverloadResolutionPriority(1)] // so that the caller argument expression is used when no message is provided
+        public static void Assert([DoesNotReturnIf(false)] bool condition, [CallerArgumentExpression(nameof(condition))] string? message = "") =>
             Assert(condition, message, string.Empty);
 
         [Conditional("DEBUG")]
+        [OverloadResolutionPriority(2)] // so that interpolation still wins out over strings
```

Comment:
Without a compiler to test it, I wasn't sure: will this be necessary? My concern was that by specifying ORP(1) on the above overload, that if someone then did `Debug.Assert(condition, $"...")`, it would end up also preferring the string overload over interpolation.

Reply:
This concern is why I said the actual change will be more complicated in the issue. I think another option is just to put a -1 on the one-arg version, to deprioritize that one. Do we only want callerargexpr in the case where the user doesn't provide a message at all, and not in addition to a message they provide?

Reply:
> But if I said "no", what would the answer be?

We'd need to add new overloads for that, then make them the highest priority, then make the single-arg the lowest priority. I think it might be best to start with the end goal via some example syntax and work backwards.

```cs
Debug.Assert(a != null);
Debug.Assert(a != null, "a wasn't null");
Debug.Assert(a != null, $"a wasn't null: {a}");
Debug.Assert(a != null, "a wasn't null", a.ToString());
Debug.Assert(a != null, $"a wasn't null ", $"a: {a}");
Debug.Assert(a != null, "a wasn't null", "a: {0}", a);
```

I believe that should hit all the existing overloads. What behaviors do we want to change?

Reply:
@stephentoub? I think we should also consider putting `EditorBrowsable.Never` on the main `Debug.Assert(bool)` method, since we hope it'll never be called. Not sure if that attribute is available in this layer though?

Reply:
Given the examples I listed above, what behaviors do we want to change?

Reply:
> Given the examples I listed above, what behaviors do we want to change?

From my perspective, it's if the user hasn't provided any message, we provide one. So just:
```C#
Debug.Assert(a != null);
```
should behave as if it were:
```C#
Debug.Assert(a != null, "a != null");
```

> I believe that should hit all the existing overloads.

And do what?

Reply:
The goal was to look at all the overloads and understand what behavior we actually wanted for each; I had imagined that you may want to _always_ include the expression, and then include user-provided messages in addition. It sounds like that's not your goal though, so here's my proposal of what we should do:

* Put `OverloadResolutionPriority(-1)` on `Debug.Assert(bool)`.
* Put `EditorBrowsable(EditorBrowsableState.Never)` on `Debug.Assert(bool)` (if possible at this layer).

Reply:
>  I had imagined that you may want to always include the expression, and then include user-provided messages in addition. It sounds like that's not your goal though

Yup, not my goal. I'm open to feedback from others saying my goals are misguided, of course.

> Put OverloadResolutionPriority(-1) on Debug.Assert(bool).

Makes sense, will do.

> Put EditorBrowsable(EditorBrowsableState.Never) on Debug.Assert(bool) (if possible at this layer).

It's of course possible. I'm concerned about possible impact on other languages, though my understanding is that both VB and F# will (eventually) catch up with C# on support for this feature. (My other concern is what impact that'll have on IntelliSense, but I _think_ the right things happen, I just need to test it.)